components/dropdown/dropdown-opener-mixin.js

```diff
@@ -0,0 +1,130 @@
+export const DropdownOpenerMixin = superclass => class extends superclass {
+
+	static get properties() {
+		return {
+			dropdownOpener: {
+				type: Boolean
+			},
+			noAutoOpen: {
+				type: Boolean,
+				reflect: true,
+				attribute: 'no-auto-open'
+			},
+			disabled: {
+				type: Boolean,
+				reflect: true
+			}
+		};
+	}
+
+	constructor() {
+		super();
+		this.dropdownOpener = true;
+		this.noAutoOpen = false;
+		this.disabled = false;
+
+		this.__onKeyPress = this.__onKeyPress.bind(this);
+		this.__onMouseUp = this.__onMouseUp.bind(this);
+	}
+
+	firstUpdated(changedProperties) {
+		super.firstUpdated(changedProperties);
+
+		this.addEventListener('d2l-dropdown-open', this.__onOpened);
+		this.addEventListener('d2l-dropdown-close', this.__onClosed);
+
+		const opener = this.getOpenerElement();
+		const content = this.__getContentElement();
+		if (!opener) {
+			return;
+		}
+		opener.setAttribute('aria-haspopup', 'true');
+		opener.addEventListener('keypress', this.__onKeyPress);
+		opener.addEventListener('mouseup', this.__onMouseUp);

```

Comment:
Should these listeners be removed at some point?

Reply:
The opener _should_ be in this elements composed sub-tree, so this _should_ get cleaned up automatically.  However, it could be in the light-DOM (as is the case when using `d2l-dropdown`), so technically the opener element is subject to any craziness user-land code might do.  For the unpredictable case, we'd need to wire up un `connectedCallback`, and remove in `disconnectedCallback`.

Reply:
This was reason I moved it from `firstUpdated` to the `connectedCallback` and added explicit `removeEventListener` calls in `disconnectedCallback`.